kratos/future/solving_strategies/builders/block_builder.h

```diff
@@ -0,0 +1,375 @@
+            rLinearSystemContainer.pEffectiveDx.swap(p_eff_dx);
+        }
+    }
+
+    void AllocateLinearSystemConstraints(
+        const DofsArrayType& rDofSet,
+        const DofsArrayType& rEffectiveDofSet,
+        const DofArrayUtilities::SlaveToMasterDofsMap& rSlaveToMasterDofsMap,
+        LinearSystemContainer<TSparseMatrixType, TSystemVectorType>& rLinearSystemContainer) override
+    {
+        // Check if there are master-slave constraints
+        if (!rSlaveToMasterDofsMap.empty()) {
+            // Fill the master-slave constraints graph
+            TSparseGraphType constraints_sparse_graph;
+            this->SetUpMasterSlaveConstraintsGraph(rDofSet, rEffectiveDofSet, rSlaveToMasterDofsMap, constraints_sparse_graph);
+
+            // Allocate the constraints arrays (note that we are using the move assignment operator in here)
+            auto p_aux_q = Kratos::make_shared<TSystemVectorType>(rDofSet.size());
+            rLinearSystemContainer.pConstraintsQ.swap(p_aux_q);
+
+            auto p_aux_T = Kratos::make_shared<TSparseMatrixType>(constraints_sparse_graph);
+            rLinearSystemContainer.pConstraintsT.swap(p_aux_T);
+        }
+    }
+
+    void BuildDirichletConstraints(
+        const DofsArrayType& rDofSet,
+        const DofsArrayType& rEffectiveDofSet,
+        LinearSystemContainer<TSparseMatrixType, TSystemVectorType>& rLinearSystemContainer) override
+    {
+    }
+
+    //FIXME: Do the RHS-only version
+    void ApplyLinearSystemConstraints(
+        const DofsArrayType& rEffectiveDofArray,
+        LinearSystemContainer<TSparseMatrixType, TSystemVectorType>& rLinearSystemContainer) override
+    {
+        // Calculate the effective LHS, RHS and solution vector
+        ApplyBlockBuildMasterSlaveConstraints(rLinearSystemContainer);
+
+        // Apply the Dirichlet BCs in a block way by leveraging the CSR matrix implementation
+        auto& r_eff_rhs = *(rLinearSystemContainer.pEffectiveRhs);
+        auto& r_eff_lhs = *(rLinearSystemContainer.pEffectiveLhs);
+        ApplyBlockBuildDirichletConditions(rEffectiveDofArray, r_eff_lhs, r_eff_rhs);
+    }
+
+    ///@}
+private:
+    ///@name Member Variables
+    ///@{
+
+    ScalingType mScalingType; // Diagonal scaling type
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief Applies the master-slave constraints
+     * This method applies the master-slave constraints following a block-type build
+     * @param rLinearSystemContainer Auxiliary container with the linear system arrays
+     */
+    void ApplyBlockBuildMasterSlaveConstraints(LinearSystemContainer<TSparseMatrixType, TSystemVectorType> &rLinearSystemContainer)
+    {
+        const auto& r_model_part = this->GetModelPart();
+        const std::size_t n_constraints = r_model_part.NumberOfMasterSlaveConstraints();
+        if (n_constraints) { //FIXME: In here we should check the number of active constraints
```

Comment:
This comments make me think this is not the final version, more like an initial version not 100% replacement of the current ones.